portality/scripts/article_cleanup_sync.py

```diff
@@ -0,0 +1,137 @@
+"""
+For each article in the DOAJ index:
+    * Checks that it has a corresponding journal, deletes it otherwise (if -d set)
+    * Ensures that the article in_doaj status is the same as the journal's
+    * Applies the journal's information to the article metadata as needed
+"""
+import esprit
+from portality import models
+from portality.util import unicode_dict
+from datetime import datetime
+from portality.core import app
+
+batch_size = 1000
+
+# Cache the Journals here, so we don't keep searching for them in the index
+journal_cache = {}
+
+
+def cleanup_articles(write_changes=False, prep_all=False):
+
+    # Connection to the ES index
+    conn = esprit.raw.make_connection(None, 'localhost', 9200, 'doaj')
+
+    write_batch = []
+    delete_batch = set()
+
+    updated_count = 0
+    same_count = 0
+    deleted_count = 0
+
+    # Scroll though all articles in the index
+    for a in esprit.tasks.scroll(conn, 'article'):
+        try:
+            article_model = models.Article(_source=a)
+
+            # Try to find journal in our cache
+            bibjson = article_model.bibjson()
+            pissns = bibjson.get_identifiers(bibjson.P_ISSN)
+            eissns = bibjson.get_identifiers(bibjson.E_ISSN)
+            allissns = list(set(pissns + eissns))
+
+            assoc_journal = None
+            for issn in allissns:
+                try:
+                    assoc_journal = journal_cache[issn]
+                    break
+                except KeyError:
+                    continue
+
+            # Cache miss; ask the article model to try to find its journal
+            if assoc_journal is None:
+                assoc_journal = article_model.get_journal()
+
+                # Add the newly found journal to our cache
+                if assoc_journal is not None:
+                    for issn in allissns:
```

Comment:
Yes, but will all the ISSNs that the article used to find the journal *actually belong* to that journal? (In edge cases where one of the ISSNs on the article record is wrong / old etc.)

Therefore should we only cache it by its own recorded ISSNs?